cedar-policy/src/api/err.rs

```diff
@@ -115,6 +116,197 @@ pub enum ReAuthorizeError {
     },
 }
 
+/// The kind of hierarchy that has transitive closure errors
+#[derive(Debug, Clone)]
+pub enum HierarchyKind {
+    /// An action hierarchy
+    Action,
+    /// An entity type hierarchy
+    EntityType,
+}
+
+impl std::fmt::Display for HierarchyKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Action => write!(f, "action"),
+            Self::EntityType => write!(f, "entity type"),
+        }
+    }
+}
+
+/// Transitive closure computation or enforcement error
+#[derive(Debug, Diagnostic, Error, Clone)]
+#[error("transitive closure computation/enforcement error on {kind} hierarchy: {message}")]
+pub struct TransitiveClosureError {
+    kind: HierarchyKind,
+    message: SmolStr,
+}
+
+impl TransitiveClosureError {
+    /// Getter of hierarchy kind
+    pub fn get_kind(&self) -> HierarchyKind {
+        self.kind.clone()
+    }
+}
+
+/// Schema constructs
+#[derive(Debug, Clone)]
+pub enum SchemaConstructKind {
+    /// schema common type
+    CommonType,
+    /// schema entity type
+    EntityType,
+    /// schema action
+    Action,
+}
+
+impl std::fmt::Display for SchemaConstructKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Action => write!(f, "action"),
+            Self::CommonType => write!(f, "common type"),
+            Self::EntityType => write!(f, "entity type"),
+        }
+    }
+}
+
+/// Undeclared schema construct error
+#[derive(Debug, Clone, Diagnostic, Error)]
+#[error("undeclared {kind}(s): {constructs:?}")]
+pub struct UndeclaredSchemaConstructError {
+    kind: SchemaConstructKind,
+    constructs: HashSet<SmolStr>,
+}
+
+impl UndeclaredSchemaConstructError {
+    /// Getter of schema construct kind
+    pub fn get_kind(&self) -> SchemaConstructKind {
+        self.kind.clone()
+    }
+
+    /// Getter of schema constructs
+    pub fn get_constructs(&self) -> impl Iterator<Item = SmolStr> + '_ {
+        self.constructs.iter().cloned()
+    }
+}
+
+/// Undeclared schema construct error
+#[derive(Debug, Clone, Diagnostic, Error)]
+#[error("duplicate {kind} `{construct}`")]
+pub struct DuplicateSchemaConstructError {
+    kind: SchemaConstructKind,
+    construct: SmolStr,
+}
+
+impl DuplicateSchemaConstructError {
+    /// Getter of schema construct kind
+    pub fn get_kind(&self) -> SchemaConstructKind {
+        self.kind.clone()
+    }
+
+    /// Getter of schema constructs
+    pub fn get_construct(&self) -> SmolStr {
+        self.construct.clone()
+    }
+}
+
+/// Cycle in action hierarchy error
+#[derive(Debug, Clone, Diagnostic, Error)]
+#[error("cycle in action hierarchy containing `{node}`")]
+pub struct CycleInActionHierarchyError {
+    node: EntityUid,
+}
+
+impl CycleInActionHierarchyError {
+    /// Getter of the action in the cycle
+    pub fn get_node_in_cycle(&self) -> SmolStr {
+        self.node.to_smolstr()
+    }
```

Comment:
Would it make sense to return the `EntityUid` here?

Reply:
I think we'd better not expose implementation details here, which include what nodes are.

Reply:
I think it's sensible for a `CycleInActionHierarchyError` to have a getter which tells users the `EntityUid` of the action that was in the cycle.  I agree we shouldn't expose implementation details about nodes, but I think returning an `EntityUid` is much more helpful to users than a string.